cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -235,6 +234,18 @@ private boolean checkJobConstraints() throws Exception {
     }
 
     private void checkAndUpdateJob(JobQueueDataset jobQueue, Job job) {
+      if (job.isToBeDeleted()) {
+        // only delete jobs that are pending trigger or pending constraint. If pending launch, the launcher will delete
+        if ((job.getState() == Job.State.PENDING_CONSTRAINT ||
+          // if pending trigger, we need to check if now - deletionTime > txTimeout. Otherwise the subscriber thread
+          // might update this job concurrently (because its tx does not see the delete flag) and cause a conflict
+          (job.getState() == Job.State.PENDING_TRIGGER
+            && System.currentTimeMillis() - job.getToBeDeletedTimestamp() > Schedulers.SUBSCRIBER_TX_TIMEOUT_MILLIS))) {
```

Comment:
`System.currentTimeMillis() - job.getToBeDeletedTimestamp()` should be a bit longer than `Schedulers.SUBSCRIBER_TX_TIMEOUT_MILLIS`. Because the `toBeDeletedTimestamp` recored in the job is earlier than the time when the transaction actually commits. There's a very small chance that the subscriber thread starts a transaction right after `toBeDeletedTimestamp` but before `markJobForDeletion` commits. So when `System.currentTimeMillis() - job.getToBeDeletedTimestamp() > Schedulers.SUBSCRIBER_TX_TIMEOUT_MILLIS`, this transaction in subscriber thread may still be in progress